src/debugAdapterDlvDap/goDlvDapDebug.ts

```diff
@@ -0,0 +1,618 @@
+		// class, but this method isn't called in 'inline' mode.
+		logger.init((e) => this.sendEvent(e));
+
+		// this debugger uses zero-based lines and columns
+		this.setDebuggerLinesStartAt1(false);
+		this.setDebuggerColumnsStartAt1(false);
+	}
+
+	protected initializeRequest(
+		response: DebugProtocol.InitializeResponse,
+		args: DebugProtocol.InitializeRequestArguments,
+		request?: DebugProtocol.Request
+	): void {
+		log('InitializeRequest');
+		response.body.supportsConfigurationDoneRequest = true;
+		response.body.supportsSetVariable = true;
+
+		// We respond to InitializeRequest here, because Delve hasn't been
+		// launched yet. Delve will start responding to DAP requests after
+		// LaunchRequest is received, which tell us how to start it.
+
+		// TODO: we could send an InitializeRequest to Delve when
+		// it launches, wait for its response and sanity check the capabilities
+		// it reports. Once DAP support in Delve is complete, this can be part
+		// of making sure that the "dlv" binary we find is sufficiently
+		// up-to-date to talk DAP with us.
+		this.sendResponse(response);
+		log('InitializeResponse');
+	}
+
+	protected launchRequest(
+		response: DebugProtocol.LaunchResponse,
+		args: LaunchRequestArguments,
+		request: DebugProtocol.Request
+	): void {
+		// Setup logger now that we have the 'trace' level passed in from
+		// LaunchRequestArguments.
+		this.logLevel =
+			args.trace === 'verbose'
+				? Logger.LogLevel.Verbose
+				: args.trace === 'log'
+					? Logger.LogLevel.Log
+					: Logger.LogLevel.Error;
+		const logPath =
+			this.logLevel !== Logger.LogLevel.Error ? path.join(os.tmpdir(), 'vscode-godlvdapdebug.txt') : undefined;
+		logger.setup(this.logLevel, logPath);
+
+		log('launchRequest');
+
+		if (!args.port) {
+			args.port = this.DEFAULT_DELVE_PORT;
+		}
+		if (!args.host) {
+			args.host = this.DEFAULT_DELVE_HOST;
+		}
+
+		// TODO: if this is a noDebug launch request, don't launch Delve;
+		// instead, run the program directly.
+
+		this.dlvClient = new DelveClient(args);
+
+		this.dlvClient.on('stdout', (str) => {
+			log('dlv stdout:', str);
+		});
+
+		this.dlvClient.on('stderr', (str) => {
+			log('dlv stderr:', str);
+		});
+
+		this.dlvClient.on('connected', () => {
+			this.dlvClient.send(request);
+		});
+
+		this.dlvClient.on('close', (rc) => {
+			if (rc !== 0) {
+				this.sendErrorResponse(
+					response,
+					3000,
+					'Failed to continue: Check the debug console for details.');
+			}
+			log('Sending TerminatedEvent as delve is closed');
+			this.sendEvent(new TerminatedEvent());
+		});
+
+		// Relay events and responses back to vscode. In the future we will
+		// add middleware here to intercept specific kinds of responses/events
+		// for special handling.
+		this.dlvClient.on('event', (event) => {
+			this.sendEvent(event);
+		});
+
+		this.dlvClient.on('response', (resp) => {
+			this.sendResponse(resp);
+		});
+	}
+
+	protected terminateRequest(
```

Comment:
Are you forwarding this and other optional unsupported requests below on purpose?

This request is not supported by vscode-go because it should never be called by the client. It is only to be supported if the initialize response sets the corresponding capability to true. The vscode-go does have a no-op fallback from the parent (although that's not always true for all optionally supported requests), but it should never actually trigger. Delve has explicit error checking just in case because it doesn't inherit from anywhere. But in the old adapter I found it useful that only actually supported and relevant requests were implemented, so I think you should consider doing this here.

Reply:
I just include them for completeness. What do you mean by the current adapter not supporting it?

AFAICS it inherits `terminateRequest` from `DebugSession` and sends a reply back to the client. This the goal of this DA is to interfere as little as possible in the DAP traffic between the editor and Delve, I figured it can forward all requests and Delve will reply "unsupported" or whatever error in case it's not supported. Otherwise we will be splitting logic between the DA and Delve which will be more difficult to follow.

Reply:
If you search for terminateRequest in the current adaptor, you will not find it. This is because capability ‘supportsTerminateRequest’ was not set to true in the InitializeResponse. Thus, the client should never issue this request and the adaptor is not expected to have any code to handle it. Same for restartRequest and many others.

It is very much in the spirit of DAP for the adaptors to be debugger-specific and "know" things about it instead of just blindly forwarding everything. If you want this adaptor to handle initialize requests and respond to the client with the what-is-supported contract instead of relying on delve to do that, then it  doesn't make sense to then forward unsupported requests to delve. This adaptor already took over that logic on initialize. I think it would be cleaner to be consistent and exclude the requests that it already claimed were unsupported.

Reply:
I spent a bit more thinking about this and I disagree.

I'm not sure the current DA's behavior here is correct. It's one thing to say "this can never happen", but another to properly error out when impossible things inevitably *do* happen. The current DA will happily return a valid `TerminateResponse`, even though it doesn't support this request. It would be much better to return an error. This is what your DAP code in Delve does correctly now - even though it doesn't say it supports the terminate request in capabilities, it will return an error with "not yet implemented" when it receives it.

This is the kind of logic that would be easiest to leave in Delve. Right now the new DA will simply forward `TerminateRequest` to Delve and then will forward back Delve's error response. I see no reason to split this logic between two code bases. It's true that the DA will have to be aware of Delve's capabilities and report them in response to the initialize request - but that's a single place in the code where this information is kept. If Delve starts supporting this request, there will be fewer changes to remember to make in the DA.

If you disagree with this, let's leave it to @hyangah to decide. Since my concern here is about long-term maintenance, I will defer to the long term maintainers of this code.

Reply:
I don't think the current debug adapter's implementation should limit the potential of the new debug adapter.
It was written many years ago before DAP reached to the current shape. 

Currently we don't set `supportsTerminateRequest`, which means this path will not be triggered. Is it better to have it for completeness (and possibly test if the feature by enabling it because giving the delve a chance to terminate debugee properly sounds like a good thing to have.)